File: typescript/chapters/ch02/src/adapters/filesystem.ts
/**
 * Node 文件系统适配器：实现 core 的 WorkspaceFileSystem。
 * 所有工具路径必须先经 safePath 从 workspace 相对路径解析为安全绝对路径，
 * 再执行 read/write/edit/glob。Node 错误在适配器边界翻译为领域错误。
 * UTF-8 使用 fatal TextDecoder，非法字节不会被静默替换。
 */
import {
  lstat,
  mkdir,
  readdir,
  readFile as readFileBytes,
  realpath,
  stat,
  writeFile as writeFileBytes,
} from "node:fs/promises";
import type { Stats } from "node:fs";
import { dirname, isAbsolute, join, relative, resolve, sep, win32 } from "node:path";
import { TextDecoder } from "node:util";

import {
  FileNotFoundError,
  FileSystemOperationError,
  InvalidFilePathError,
  InvalidUtf8Error,
  TextNotFoundError,
  WorkspacePathError,
} from "../core/filesystem.js";
import type { WorkspaceFileSystem } from "../core/filesystem.js";

// Windows 保留设备名和严格 UTF-8 解码器都属于工作区安全边界。
// Win32 设备文件名集合；即使带扩展名也不能作为普通工作区路径组件。
const WINDOWS_DEVICE_NAMES = new Set(["AUX", "CLOCK$", "CON", "CONIN$", "CONOUT$", "NUL", "PRN"]);
// fatal 模式避免无效字节被替换字符掩盖，读工具必须报告编码错误。
const UTF8_DECODER = new TextDecoder("utf-8", { fatal: true });

// 文件系统适配器把 Node 错误归一为工具层可映射的领域错误。
// 从未知异常中提取 Node 的 code，不能信任任何异常对象一定具有该字段。
function errorCode(error: unknown): string | undefined {
  if (typeof error !== "object" || error === null) {
    return undefined;
  }
  const code = Reflect.get(error, "code");
  return typeof code === "string" ? code : undefined;
}

// 将 Node 的宽松异常映射为核心文件系统契约的稳定错误类型。
function translateFileSystemError(error: unknown): Error {
  // 已知领域错误原样保留，其余 Node 错误映射为稳定的工具层错误类别。
  if (
    error instanceof WorkspacePathError ||
    error instanceof TextNotFoundError ||
    error instanceof InvalidUtf8Error ||
    error instanceof FileNotFoundError ||
    error instanceof InvalidFilePathError ||
    error instanceof FileSystemOperationError ||
    error instanceof RangeError
  ) {
    return error;
  }
  const code = errorCode(error);
  if (code === "ENOENT") {
    return new FileNotFoundError("File or directory was not found");
  }
  if (code === "EISDIR" || code === "ENOTDIR") {
    return new InvalidFilePathError("Path has the wrong file type");
  }
  if (code === undefined) {
    return error instanceof Error ? error : new Error("Non-error value thrown by file system code");
  }
  return new FileSystemOperationError("File system operation failed");
}

// 判断候选绝对路径是否仍在根目录内，兼容 Windows 与 POSIX 分隔符语义。
function isInside(root: string, candidate: string): boolean {
  // relative 结果既非父目录也非绝对路径，才表示 candidate 仍在 root 内。
  const child = relative(root, candidate);
  return (
    child.length === 0 || (child !== ".." && !child.startsWith(`..${sep}`) && !isAbsolute(child))
  );
}

// 拒绝 Windows 非法字符、尾随空格/点以及设备名，防止跨平台路径语义偏差。
function isWindowsReservedComponent(component: string): boolean {
  // 同时拒绝 Win32 非法字符、尾随点/空格和设备文件名。
  if (component.endsWith(" ") || component.endsWith(".")) {
    return true;
  }
  if (
    [...component].some((character) => {
      const code = character.codePointAt(0);
      return (code !== undefined && code < 32) || '<>:"|*?'.includes(character);
    })
  ) {
    return true;
  }
  const stem = component.split(".", 1)[0]?.replace(/ +$/u, "").toUpperCase();
  if (stem === undefined) {
    return false;
  }
  if (WINDOWS_DEVICE_NAMES.has(stem)) {
    return true;
  }
  return (
    stem.length === 4 &&
    (stem.startsWith("COM") || stem.startsWith("LPT")) &&
    "123456789¹²³".includes(stem[3] ?? "")
  );
}

// 将用户路径或 glob 拆成安全相对组件；allowWildcards 只为 glob 的模式字符放行。
function relativeParts(value: string, label: string, allowWildcards: boolean): string[] {
  // 同时拒绝 POSIX、Win32 和盘符绝对路径，所有工具路径必须相对工作区。
  if (value.length === 0) {
    throw new WorkspacePathError(`${label} must not be empty`);
  }
  if (value.includes("\0")) {
    throw new WorkspacePathError(`${label} contains a null byte`);
  }
  const normalized = value.replaceAll("\\", "/");
  if (
    isAbsolute(value) ||
    win32.isAbsolute(value) ||
    /^[A-Za-z]:/u.test(value) ||
    normalized.startsWith("/")
  ) {
    throw new WorkspacePathError(
      `${label} must be relative; absolute paths are rejected: ${value}`,
    );
  }
  const parts = normalized.split("/").filter((part) => part.length > 0 && part !== ".");
  if (parts.includes("..")) {
    throw new WorkspacePathError(`${label} must not contain parent segments: ${value}`);
  }
  for (const part of parts) {
    if (!allowWildcards && isWindowsReservedComponent(part)) {
      throw new WorkspacePathError(`${label} contains a reserved Windows path component: ${part}`);
    }
    if (
      [...part].some((character) => {
        const code = character.codePointAt(0);
        return (code !== undefined && code < 32) || '<>:"|'.includes(character);
      })
    ) {
      throw new WorkspacePathError(`${label} contains a reserved Windows path component: ${part}`);
    }
  }
  return parts;
}

// 解析并验证工作区真实路径，作为符号链接检查所依据的可信根。
async function workspaceRoot(workspace: string): Promise<string> {
  // 真实路径是后续链接逃逸检查的可信根，而不是调用方传入的词法路径。
  const root = await realpath(workspace);
  const information = await stat(root);
  if (!information.isDirectory()) {
    throw new InvalidFilePathError(`Workspace is not a directory: ${workspace}`);
  }
  return root;
}

// 向上查找目标最接近的已存在父目录，以便写入新文件前也能验证真实路径边界。
async function resolvedExistingParent(
  root: string,
  target: string,
): Promise<{ physical: string; lexical: string }> {
  // 写入目标可能尚不存在，向上寻找首个可 realpath 的父目录即可验证链接边界。
  let current = target;
  while (true) {
    try {
      return { physical: await realpath(current), lexical: current };
    } catch (error) {
      if (errorCode(error) !== "ENOENT" || current === root) {
        throw error;
      }
      current = dirname(current);
    }
  }
}

// 以词法检查和 realpath 检查双重解析相对路径，阻止 junction 与符号链接逃逸。
export async function safePath(workspace: string, relativePath: string): Promise<string> {
  // 词法检查后再解析现有父目录真实路径，双重防护链接逃逸。
  try {
    const root = await workspaceRoot(workspace);
    const parts = relativeParts(relativePath, "path", false);
    const target = resolve(root, ...parts);
    if (!isInside(root, target)) {
      throw new WorkspacePathError(`Path escapes workspace: ${relativePath}`);
    }
    // 现有父路径必须解析真实位置，防止 junction 或符号链接把目标带出工作区。
    const existing = await resolvedExistingParent(root, target);
    const resolved = resolve(existing.physical, relative(existing.lexical, target));
    if (!isInside(root, resolved)) {
      throw new WorkspacePathError(`Path escapes workspace: ${relativePath}`);
    }
    return resolved;
  } catch (error) {
    throw translateFileSystemError(error);
  }
}

// 用共享严格解码器读取文本，文件含非法 UTF-8 时返回领域错误而非替换字符。
function decodeUtf8(bytes: Uint8Array, relativePath: string): string {
  try {
    return UTF8_DECODER.decode(bytes);
  } catch {
    throw new InvalidUtf8Error(`File is not valid UTF-8: ${relativePath}`);
  }
}

// 统一换行符后按行截断，输出不依赖宿主系统的行尾格式。
// 尾随空行不计作内容行，保证 read_file 的 limit 在各平台一致。
function splitLines(text: string): string[] {
  if (text.length === 0) {
    return [];
  }
  const lines = text.split(/\r\n|\r|\n/u);
  if (lines.at(-1) === "") {
    lines.pop();
  }
  return lines;
}

// 将约定的 glob 子集编译为路径级正则；不把未实现模式静默交给宿主 shell。
function globRegex(pattern: string): RegExp {
  // 仅实现工具约定的 glob 子集，匹配对象始终是标准化的相对路径。
  let source = "^";
  for (let index = 0; index < pattern.length; index += 1) {
    const character = pattern[index];
    if (character === undefined) {
      continue;
    }
    if (character === "*" && pattern[index + 1] === "*") {
      index += 1;
      if (pattern[index + 1] === "/") {
        index += 1;
        source += "(?:.*/)?";
      } else {
        source += ".*";
      }
    } else if (character === "*") {
      source += "[^/]*";
    } else if (character === "?") {
      source += "[^/]";
    } else if (character === "[") {
      const end = pattern.indexOf("]", index + 1);
      if (end === -1) {
        source += "\\[";
      } else {
        const content = pattern.slice(index + 1, end);
        source += content.startsWith("!") ? `[^${content.slice(1)}]` : `[${content}]`;
        index = end;
      }
    } else {
      source += character === "/" ? "/" : character.replace(/[\\^$+?.()|{}]/gu, "\\$&");
    }
  }
  return new RegExp(`${source}$`, "u");
}

// 提取第一个通配符前的确定目录，以缩小 glob 扫描范围并先验证该路径。
async function literalPrefix(root: string, parts: readonly string[]): Promise<readonly string[]> {
  // 从第一个通配符起停止，减少 glob 遍历起点并先验证确定的目录前缀。
  const literal: string[] = [];
  for (const part of parts) {
    if (/[?*[\]]/u.test(part)) {
      break;
    }
    literal.push(part);
  }
  if (literal.length > 0) {
    await safePath(root, literal.join(sep));
  }
  return literal;
}

// Node fs/promises 对 WorkspaceFileSystem 的实现；所有公开操作都通过 safePath 保持工作区边界。
export class NodeWorkspaceFileSystem implements WorkspaceFileSystem {
  // 所有 I/O 都经 safePath，不能绕过工作区边界直接操作宿主文件。
  // 读取后统一行尾，limit 只影响返回内容，不影响磁盘文件。
  async readFile(workspace: string, relativePath: string, limit?: number): Promise<string> {
    try {
      if (limit !== undefined && (!Number.isInteger(limit) || limit <= 0)) {
        throw new RangeError("limit must be a positive integer");
      }
      const target = await safePath(workspace, relativePath);
      const text = decodeUtf8(await readFileBytes(target), relativePath);
      const lines = splitLines(text);
      if (limit !== undefined && limit < lines.length) {
        return [...lines.slice(0, limit), `... (${lines.length - limit} more lines)`].join("\n");
      }
      return lines.join("\n");
    } catch (error) {
      throw translateFileSystemError(error);
    }
  }

  // 在安全目标的父目录内创建缺失目录并写入 UTF-8 字节，返回写入大小。
  async writeFile(workspace: string, relativePath: string, content: string): Promise<number> {
    try {
      const target = await safePath(workspace, relativePath);
      // 仅在已确认安全的目标父目录中递归创建目录。
      const bytes = Buffer.from(content, "utf8");
      await mkdir(dirname(target), { recursive: true });
      await writeFileBytes(target, bytes);
      return bytes.byteLength;
    } catch (error) {
      throw translateFileSystemError(error);
    }
  }

  // 读取、查找并只替换第一次精确文本；找不到时抛出 TextNotFoundError。
  async editFile(
    workspace: string,
    relativePath: string,
    oldText: string,
    newText: string,
  ): Promise<void> {
    // 精确文本替换只修改首次命中，调用方可据此构造可预测的编辑操作。
    try {
      if (oldText.length === 0) {
        throw new RangeError("old_text must not be empty");
      }
      const target = await safePath(workspace, relativePath);
      const currentBytes = await readFileBytes(target);
      const current = decodeUtf8(currentBytes, relativePath);
      const index = current.indexOf(oldText);
      if (index === -1) {
        throw new TextNotFoundError(`Exact text not found in ${relativePath}`);
      }
      const updated = `${current.slice(0, index)}${newText}${current.slice(index + oldText.length)}`;
      await writeFileBytes(target, Buffer.from(updated, "utf8"));
    } catch (error) {
      throw translateFileSystemError(error);
    }
  }

  // 在安全工作区内遍历目录并返回稳定、去重、使用 / 分隔的相对匹配结果。
  async globFiles(workspace: string, pattern: string): Promise<readonly string[]> {
    try {
      const root = await workspaceRoot(workspace);
      const parts = relativeParts(pattern, "glob pattern", true);
      const prefix = await literalPrefix(root, parts);
      const normalizedPattern = parts.length === 0 ? "." : parts.join("/");
      const matcher = globRegex(normalizedPattern);
      const start = resolve(root, ...prefix);
      let startInformation: Stats;
      try {
        startInformation = await lstat(start);
      } catch (error) {
        if (errorCode(error) === "ENOENT") {
          return Object.freeze([]);
        }
        throw error;
      }
      if (prefix.length === parts.length) {
        // 无通配符时无需扫描目录，直接按规范化相对路径返回。
        return Object.freeze(matcher.test(normalizedPattern) ? [normalizedPattern] : []);
      }
      if (!startInformation.isDirectory() || startInformation.isSymbolicLink()) {
        return Object.freeze([]);
      }
      const results: string[] = [];
      const pending = [start];
      // Dirent.isDirectory() 不会跟随符号链接目录，避免递归越界和链接环。
      while (pending.length > 0) {
        const directory = pending.pop();
        if (directory === undefined) {
          continue;
        }
        for (const entry of await readdir(directory, { withFileTypes: true })) {
          const target = join(directory, entry.name);
          const relativePath = relative(root, target).split(sep).join("/");
          if (matcher.test(relativePath)) {
            const resolved = await safePath(root, relativePath);
            if (!isInside(root, resolved)) {
              throw new WorkspacePathError(`Glob match escapes workspace: ${relativePath}`);
            }
            results.push(relativePath);
          }
          if (entry.isDirectory()) {
            pending.push(target);
          }
        }
      }
      return Object.freeze([...new Set(results)].sort());
    } catch (error) {
      throw translateFileSystemError(error);
    }
  }
}

File: python/ch07_agent/agent_ch07/adapters/filesystem.py
"""本地文件系统适配器。

这是什么：
    本地文件系统的适配器实现，封装 pathlib/os 的细节并提供路径安全边界。

Java 类比：
    类似 Java 的基础设施适配器：把 `Files`/`Paths` 的细节藏起来，
    并在每次操作前检查工作区边界、符号链接和 Windows 保留路径名。

为什么需要：
    - 每次操作前检查工作区边界、符号链接和 Windows 保留路径名
    - 统一错误处理，将 OSError 转换为领域异常
    - 核心层通过 WorkspaceFileSystem 接口调用，不直接依赖 pathlib
    - 提供路径安全函数供其他模块（如 Skill）复用

这层相当于 Java 的基础设施适配器：把 `pathlib`/`os` 的细节藏起来，
并在每次操作前检查工作区边界、符号链接和 Windows 保留路径名。
"""

from __future__ import annotations

import fnmatch
import re
from pathlib import Path

from ..core.filesystem import (
    FileNotFoundError,
    FileSystemOperationError,
    InvalidFilePathError,
    InvalidUtf8Error,
    TextNotFoundError,
    WorkspaceFileSystem,
    WorkspacePathError,
)

_WINDOWS_DEVICE_NAMES = {"AUX", "CLOCK$", "CON", "CONIN$", "CONOUT$", "NUL", "PRN"}


def _translate_os_error(error: OSError) -> Exception:
    """把 Python 操作系统异常转换成稳定的领域异常。"""
    if isinstance(error, FileNotFoundError):
        return error
    if getattr(error, "errno", None) == 2:
        return FileNotFoundError("文件或目录不存在")
    if getattr(error, "errno", None) in {20, 21}:
        return InvalidFilePathError("路径指向了错误的文件类型")
    return FileSystemOperationError("文件系统操作失败")


def _is_windows_reserved(component: str) -> bool:
    """拒绝 Windows 设备名、非法字符和尾随空格/点。"""
    if component.endswith((" ", ".")):
        return True
    if any(ord(char) < 32 or char in '<>:"|*?' for char in component):
        return True
    stem = component.split(".", 1)[0].rstrip(" ").upper()
    if stem in _WINDOWS_DEVICE_NAMES:
        return True
    return len(stem) == 4 and stem[:3] in {"COM", "LPT"} and stem[3] in "123456789"


def is_windows_reserved_component(component: str) -> bool:
    """公开给其他路径功能复用的 Windows 保留组件判断。

    这是什么：
        检查路径组件是否是 Windows 保留名称的公共函数。

    Java 类比：
        类似把内部校验方法提炼成 package-private/public utility，避免
        Skill 和文件工具各自维护一份容易漂移的规则。

    为什么需要：
        - 防止使用 Windows 设备名（NUL、CON、PRN、AUX 等）导致系统异常
        - 拒绝尾随空格/点、非法字符（<>:"|*?）
        - 供 Skill 路径校验等其他模块复用，保证规则一致性
    """
    return _is_windows_reserved(component)


def _relative_parts(value: str, label: str, allow_wildcards: bool = False) -> list[str]:
    """把用户路径拆成安全相对组件；绝对路径和 `..` 永远拒绝。"""
    if not value:
        raise WorkspacePathError(f"{label} 不能为空")
    normalized = value.replace("\\", "/")
    if "\x00" in normalized or normalized.startswith("/") or re.match(r"^[A-Za-z]:", normalized):
        raise WorkspacePathError(f"{label} 必须是相对路径，不能使用绝对路径")
    parts = [part for part in normalized.split("/") if part not in {"", "."}]
    if ".." in parts:
        raise WorkspacePathError(f"{label} 不能包含父目录片段 ..")
    for part in parts:
        if not allow_wildcards and _is_windows_reserved(part):
            raise WorkspacePathError(f"{label} 包含 Windows 保留路径组件: {part}")
        if any(ord(char) < 32 or char in '<>:"|' for char in part):
            raise WorkspacePathError(f"{label} 包含非法路径组件: {part}")
    return parts


def _workspace_root(workspace: str) -> Path:
    """取得工作区真实目录，并拒绝把文件当成工作区。"""
    try:
        root = Path(workspace).resolve(strict=True)
    except OSError as error:
        raise _translate_os_error(error) from error
    if not root.is_dir():
        raise InvalidFilePathError(f"工作区不是目录: {workspace}")
    return root


def safe_path(workspace: str, relative_path: str) -> Path:
    """解析工作区相对路径，并检查词法路径和真实路径都没有逃逸。"""
    root = _workspace_root(workspace)
    parts = _relative_parts(relative_path, "路径")
    target = root.joinpath(*parts)
    try:
        existing = target
        while not existing.exists() and existing != root:
            existing = existing.parent
        physical_parent = existing.resolve(strict=True)
    except OSError as error:
        raise _translate_os_error(error) from error
    candidate = physical_parent.joinpath(*target.relative_to(existing).parts).resolve()
    try:
        candidate.relative_to(root)
    except ValueError as error:
        raise WorkspacePathError(f"路径逃逸工作区: {relative_path}") from error
    return candidate


class LocalWorkspaceFileSystem(WorkspaceFileSystem):
    """基于 pathlib 的真实工作区实现。

    这是什么：
        实现 WorkspaceFileSystem 接口的本地文件系统适配器。

    Java 类比：
        class LocalWorkspaceFileSystem implements WorkspaceFileSystem
        类似适配器模式，将操作系统 API 适配到内部接口。

    为什么需要：
        - 实现核心层定义的文件系统接口（读、写、编辑、glob）
        - 每次操作前检查路径安全（工作区边界、符号链接）
        - 统一错误处理，转换为领域异常
        - 测试时可以注入 Fake，不需要真实文件系统
    """

    def is_path_within_workspace(self, workspace: str, relative_path: str) -> bool:
        """供权限策略在写入前检查真实路径边界。"""
        try:
            safe_path(workspace, relative_path)
            return True
        except WorkspacePathError:
            return False

    def read_file(self, workspace: str, relative_path: str, limit: int | None = None) -> str:
        """严格读取文本；超出 limit 时追加剩余行数提示。"""
        if limit is not None and (
            not isinstance(limit, int) or isinstance(limit, bool) or limit <= 0
        ):
            raise ValueError("limit 必须是正整数")
        try:
            target = safe_path(workspace, relative_path)
            if not target.exists():
                raise FileNotFoundError(f"文件不存在: {relative_path}")
            if not target.is_file():
                raise InvalidFilePathError(f"路径不是文件: {relative_path}")
            try:
                text = target.read_bytes().decode("utf-8")
            except UnicodeDecodeError as error:
                raise InvalidUtf8Error(f"文件不是合法 UTF-8: {relative_path}") from error
        except (WorkspacePathError, InvalidFilePathError, InvalidUtf8Error, FileNotFoundError):
            raise
        except OSError as error:
            raise _translate_os_error(error) from error
        lines = text.replace("\r\n", "\n").replace("\r", "\n").splitlines()
        if limit is not None and limit < len(lines):
            return "\n".join([*lines[:limit], f"... ({len(lines) - limit} more lines)"])
        return "\n".join(lines)

    def write_file(self, workspace: str, relative_path: str, content: str) -> int:
        """创建父目录后写入 UTF-8 字节，并返回字节数。"""
        try:
            target = safe_path(workspace, relative_path)
            target.parent.mkdir(parents=True, exist_ok=True)
            data = content.encode("utf-8")
            target.write_bytes(data)
            return len(data)
        except WorkspacePathError:
            raise
        except OSError as error:
            raise _translate_os_error(error) from error

    def edit_file(self, workspace: str, relative_path: str, old_text: str, new_text: str) -> None:
        """替换第一次精确文本；旧文本不存在时保证不写盘。"""
        if not old_text:
            raise ValueError("old_text 不能为空")
        target = safe_path(workspace, relative_path)
        try:
            current = target.read_bytes().decode("utf-8")
        except UnicodeDecodeError as error:
            raise InvalidUtf8Error(f"文件不是合法 UTF-8: {relative_path}") from error
        except OSError as error:
            raise _translate_os_error(error) from error
        if old_text not in current:
            raise TextNotFoundError(f"文件中找不到精确文本: {relative_path}")
        updated = current.replace(old_text, new_text, 1)
        try:
            target.write_bytes(updated.encode("utf-8"))
        except OSError as error:
            raise _translate_os_error(error) from error

    def glob_files(self, workspace: str, pattern: str) -> tuple[str, ...]:
        """遍历工作区并返回稳定排序的 POSIX 风格相对路径。"""
        root = _workspace_root(workspace)
        parts = _relative_parts(pattern, "glob 模式", allow_wildcards=True)
        normalized = "/".join(parts) if parts else "."
        # 第一个通配符之前是确定路径，必须先走 safe_path。
        # 例如 escape/*.txt 中 escape 若是指向工作区外的链接，应明确报错，
        # 不能因为遍历时跳过链接就伪装成“没有匹配”。
        literal_parts: list[str] = []
        for part in parts:
            if any(char in part for char in "*?["):
                break
            literal_parts.append(part)
        if literal_parts:
            safe_path(workspace, "/".join(literal_parts))
        results: list[str] = []
        for path in root.rglob("*"):
            if path.is_symlink():
                # 不跟随符号链接，避免 glob 递归到工作区外。
                continue
            relative = path.relative_to(root).as_posix()
            matches = fnmatch.fnmatchcase(relative, normalized)
            # Python fnmatch 不把 `**/` 解释为“零层或多层目录”，
            # 因此额外尝试去掉前缀，保证 `**/*.txt` 也匹配根目录文件。
            if normalized.startswith("**/"):
                matches = matches or fnmatch.fnmatchcase(relative, normalized[3:])
            if path.is_file() and matches:
                safe_path(workspace, relative)
                results.append(relative)
        if not results and not any(char in normalized for char in "*?["):
            candidate = root / normalized
            if candidate.exists() and candidate.is_file():
                results.append(normalized)
        return tuple(sorted(set(results)))
